casper/src/main/scala/io/casperlabs/casper/Estimator.scala

```
@@ -95,25 +96,23 @@ object Estimator {
       .map(_.collect { case (v, Some(m)) => v -> m })
 
   /**
-    * Finds the latest message of validator `v` voting for block `b`, if any.
+    * Finds the main child of `b` that `v` votes for, if any.
     */
-  private def latestMessageForBlock[F[_]: MonadThrowable](
+  private def voteForChild[F[_]: MonadThrowable](
       b: BlockHash,
       v: Validator,
       latestMessage: Message,
       dag: DagRepresentation[F]
-  ): F[Option[Message]] =
-    DagOperations
-      .swimlaneV[F](v, latestMessage, dag)
-      .findF(isMainAncestor(b, _, dag))
-
-  private def isMainAncestor[F[_]: MonadThrowable](
-      smallerRank: BlockHash,
-      largerRank: Message,
-      dag: DagRepresentation[F]
-  ): F[Boolean] = dag.lookup(smallerRank) flatMap {
-    case None          => false.pure[F]
-    case Some(message) => DagOperations.isMainAncestor(message, largerRank, dag)
+  ): F[Option[Message]] = dag.lookup(b) flatMap {
+    case None => none[Message].pure[F]
```

I think this should be an error. First, so we don't mistake it with the case when `latestMessage` is an empty vote in that b-game. Second, we already should have `b` in our DAG.

---

Would you like me to add `lookupUnsafe` as part of this PR? Or shall we ticket it and do it separately?

---

Let's ticket it separately. Can you change your code here to fail on the missing dependency though? It will be very cumbersome later to track down all places where we use `dag.lookup` to replace with this `lookupUnsafe`.

---

No, I just made a general comment that instead of debating every time if it should be returning `none` or an error, while we have quite a few occurrences of `.map(lookup).map(_.flatten)`, which is the same, we should, at some point, add a method that makes both of these unneeded. 

The codebase is full of these. It's up to you if you want to do it now, I'd not ask you go fix everything, but hopefully if someone adds it at some point we can gradually (or at once as technical debt service) adopt it.